nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-jetty/src/main/java/org/apache/nifi/web/server/JettyServer.java

```diff
@@ -602,6 +599,28 @@ private WebAppContext loadWar(final File warFile, final String contextPath, fina
         return webappContext;
     }
 
+    private void addHTTPHeaders(WebAppContext webappContext) {
+        // Add a filter to set the X-Frame-Options header
+        FilterHolder xfoFilter = new FilterHolder(new XFrameOptionsFilter());
+        xfoFilter.setName(XFrameOptionsFilter.class.getSimpleName());
+        webappContext.addFilter(xfoFilter, "/*", EnumSet.allOf(DispatcherType.class));
+
+        // Add a filter to set the Content Security Policy frame-ancestors directive
+        FilterHolder cspFilter = new FilterHolder(new ContentSecurityPolicyFilter());
+        cspFilter.setName(ContentSecurityPolicyFilter.class.getSimpleName());
+        webappContext.addFilter(cspFilter, "/*", EnumSet.allOf(DispatcherType.class));
+
+        // Add a filter to set the HSTS header
```

Comment:
One more thought, and I don't know the answer on this. I like the separation of concerns where each `Filter` implementation has a specific role. However, in each case, it's just adding an HTTP header to the response. Do we have any idea of the performance impact of chaining four distinct filter implementations to do this, rather than just using a single filter which adds all 4 headers?

Reply:
Talked to @mcgilman and we believe the performance impact of using multiple filters will be negligible. Don't worry about this.